flocker/acceptance/endtoend/test_dataset.py

```diff
@@ -102,3 +109,69 @@ def not_exists():
             return deleted
         created.addCallback(delete_dataset)
         return created
+
+    @require_moving_backend
+    @run_test_with(async_runner(timeout=timedelta(minutes=6)))
+    @require_cluster(2)
+    def test_dataset_move_from_dead_node(self, cluster):
+        """
+        A dataset can be moved from one node to another.
+
+        All attributes, including the maximum size, are preserved.
+        """
+        api = get_backend_api(self, cluster.cluster_uuid)
+        if not ICloudAPI.providedBy(api):
+            raise SkipTest(
+                "Backend doesn't support ICloudAPI; therefore it might support"
+                " moving from dead node but as first pass we assume it "
+                "doesn't.")
+
+        # Find a node which is not running the control service.
+        # If the control node is shut down we won't be able to move anything!
+        node = list(node for node in cluster.nodes
+                    if node.public_address !=
+                    cluster.control_node.public_address)[0]
+        other_node = list(other_node for other_node in cluster.nodes
+                          if other_node != node)[0]
+        waiting_for_create = create_dataset(self, cluster, node=node)
+
+        def startup_node(node_id):
+            api.start_node(node_id)
+            # Give node some minimal amount of time to boot so next test
+            # is happier:
+            sleep(20)
+
+        # Once created, shut down origin node and then request to move the
+        # dataset to node2:
+        def shutdown(dataset):
+            live_node_ids = set(api.list_live_nodes())
+            d = node.shutdown()
+            # Wait for shutdown to be far enough long that node is down:
+            d.addCallback(
+                lambda _:
+                loop_until(reactor, lambda:
+                           set(api.list_live_nodes()) != live_node_ids))
+            # Schedule node start up:
+            d.addCallback(
+                lambda _: self.addCleanup(
+                    startup_node,
+                    (live_node_ids - set(api.list_live_nodes())).pop()))
```

Comment:
:grey_question: It's sorta goofy that you have to do this to find the cloud node_id. Just reading the code it'd be cool if this could just be node.cloud_id or something, but it looks like we never propagate compute_instance_id() outside of the dataset agent so that's probably impossible.

No action required, but if you have some clever idea of how to make this code more obvious, I was just noting that it seemed slightly convoluted.